Client/multiplayer_sa/CMultiplayerSA.cpp

```diff
@@ -578,6 +578,76 @@ CMultiplayerSA::CMultiplayerSA()
     m_dwLastStaticAnimID = eAnimID::ANIM_ID_WALK;
 }
 
+namespace GrainEffect
+{
+
+static BYTE ucGrainEnabled = FALSE;
+static DWORD dwGrainStrength = 1.0f;
+
+struct MasterModifier
+{
+    static float fMultiplier;
+
+    static int ApplyEffect(BYTE ucLevel, BYTE ucUpdate)
+    {
+        return ((int(__cdecl*)(char, char))0x7037C0)(ucLevel * fMultiplier, ucUpdate);
+    }
+};
+
+struct InfraredModifier
+{
+    static float fMultiplier;
+
+    static int ApplyEffect(BYTE ucLevel, BYTE ucUpdate)
+    {
+        return MasterModifier::ApplyEffect(ucLevel * fMultiplier, ucUpdate);
+    }
+};
+
+struct NightModifier
+{
+    static float fMultiplier;
+
+    static int ApplyEffect(BYTE ucLevel, BYTE ucUpdate)
+    {
+        return MasterModifier::ApplyEffect(ucLevel * fMultiplier, ucUpdate);
+    }
+};
+
+struct RainModifier
+{
+    static float fMultiplier;
+
+    static int ApplyEffect(BYTE ucLevel, BYTE ucUpdate)
+    {
+        return MasterModifier::ApplyEffect(ucLevel * fMultiplier, ucUpdate);
+    }
+};
+
+struct OverlayModifier
+{
+    static float fMultiplier;
+
+    static int ApplyEffect(BYTE ucLevel, BYTE ucUpdate)
+    {
+        return MasterModifier::ApplyEffect(ucLevel * fMultiplier, ucUpdate);
+    }
+};
+
+float MasterModifier::fMultiplier = 1.0f;
+float InfraredModifier::fMultiplier = 1.0f;
+float NightModifier::fMultiplier = 1.0f;
+float RainModifier::fMultiplier = 1.0f;
+float OverlayModifier::fMultiplier = 1.0f;
+
+}
+
+template <typename T>
+int _cdecl HOOK_GrainEffect(char a1, char a2)
+{
+    return T::ApplyEffect(a1, a2);
+}
```

Comment:
Is there some special reason for this template? It seems a bit pointless, since you should be able to just do
```
HookInstallCall(0x123456, (DWORD)GrainEffect::OverlayModifier::ApplyEffect);
```
instead